File: srm-744-div-2/third.cpp
#include <vector>
#include <cmath>
#include <algorithm>

using namespace std;


class ModularQuadrant {
public:

    long long sum(int r1, int r2, int c1, int c2) {
        return calcRect(r2, c2) - calcRect(r1 - 1, c2) - calcRect(r2, c1 - 1) + calcRect(r1 - 1, c1 - 1);
    }

    // If we have sequence like this x--x--x--x--x--... of length n, how many times does x appear?
    long long numAppear(long long n) {
        if (n <= 0) return 0;
        return n / 3 + (n % 3 ? 1 : 0);
    }

    // Sum of n consecutive numbers, starting with 'start' and step of 'step'.
    long long sumNum(long long start, long long step, long long n) {
        if (n <= 0) return 0;
        return start * n + step * (n - 1) * n / 2;
    }

    // Calculates sum for rectangle from (0, 0) to (r, c).
    long long calcRect(const long long r, const long long c) {
        if (r > c) return calcRect(c, r);
        if (r < 0 || c < 0) return 0;

        // Calc rectangle part.
        const long long w = c - r; // width of rectangle part.
        const long long h = r + 1;
        const long long x = h % 3;

        const long long sumRectPart = h * (x * numAppear(w) +
                                           (x + 1) % 3 * numAppear(w - 1) +
                                           (x + 2) % 3 * numAppear(w - 2));


        // Calc triangle part.
        const long long a = r + 1;
        const long long sumTrngPart = 0 + 1 * sumNum(2, 3, numAppear(a - 1)) + 2 * sumNum(3, 3, numAppear(a - 2));
        const long long trngDiag = numAppear(a - 1) + 2 * numAppear(a - 2); // Diagonal.

        return sumRectPart + sumTrngPart * 2 - trngDiag;
    }
};

File: srm-744-div-2/third.hs
module ModularQuadrant (sum') where

sum' :: Int -> Int -> Int -> Int -> Integer
sum' r1 r2 c1 c2 = sumRect (r2, c2) - sumRect (r1 - 1, c2) - sumRect (r2, c1 - 1) + sumRect (r1 - 1, c1 - 1)

numAppear :: (Integral a) => a -> a
numAppear n | n <= 0 = 0
            | otherwise = n `div` 3 + (if n `mod` 3 == 0 then 0 else 1)

sumNum :: (Integral a) => a -> a -> a -> a
sumNum start step n | n <= 0 = 0
                    | otherwise = start * n + step * (n - 1) * n `div` 2

sumRect :: (Int, Int) -> Integer
sumRect (r, c)
  | r < 0 || c < 0 = 0
  | r > c = sumRect (c, r)
  | otherwise = sumRectPart + 2 * sumTrngPart - sumTrngDiag
  where
    (r', c') = (fromIntegral r :: Integer, fromIntegral c :: Integer)
    sumRectPart = (* h) $ sum $ map (\d -> ((h + d) `mod` 3) * numAppear(w - d)) [0..2]
      where (h, w) = (r' + 1, c' - r')
    sumTrngPart = sumNum 2 3 (numAppear r') + 2 * (sumNum 3 3 (numAppear (r' - 1)))
    sumTrngDiag = numAppear r' + 2 * (numAppear (r' - 1))
